File: apps/mobile/ios/Happy/Services/VoiceService.swift
//
//  VoiceService.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import AVFoundation
import Combine
import Foundation

// MARK: - VoiceServiceProtocol

/// Protocol for text-to-speech services, enabling dependency injection and testing.
protocol VoiceServiceProtocol: AnyObject {
    /// Current playback state.
    var playbackState: VoicePlaybackState { get }

    /// Publisher for playback state changes.
    var playbackStatePublisher: AnyPublisher<VoicePlaybackState, Never> { get }

    /// Publisher for errors.
    var errorPublisher: AnyPublisher<VoiceServiceError, Never> { get }

    /// Speak the given text aloud.
    func speak(_ text: String) async

    /// Stop any current speech.
    func stop()

    /// Pause current speech.
    func pause()

    /// Resume paused speech.
    func resume()

    /// Update the voice settings.
    func updateSettings(_ settings: VoiceSettings)

    /// Check if the current provider is available and configured.
    func isProviderAvailable() -> Bool
}

// MARK: - Playback State

/// Represents the current state of voice playback.
enum VoicePlaybackState: Equatable {
    /// No speech in progress.
    case idle

    /// Speech is in progress.
    case speaking

    /// Speech is paused.
    case paused

    /// Loading audio from ElevenLabs API.
    case loading
}

// MARK: - VoiceServiceError

/// Errors that can occur during voice operations.
enum VoiceServiceError: LocalizedError, Equatable {
    /// ElevenLabs API key is not configured.
    case apiKeyMissing

    /// Network request to ElevenLabs failed.
    case networkError(String)

    /// ElevenLabs API returned an error.
    case apiError(statusCode: Int, message: String)

    /// Failed to decode audio data.
    case audioDecodingFailed

    /// Audio session configuration failed.
    case audioSessionError(String)

    /// The selected voice is not available.
    case voiceNotAvailable(String)

    /// Speech synthesis failed.
    case synthesisFailed(String)

    var errorDescription: String? {
        switch self {
        case .apiKeyMissing:
            return "ElevenLabs API key is not configured. Please add it in Voice Settings."
        case .networkError(let message):
            return "Network error: \(message)"
        case .apiError(let statusCode, let message):
            return "ElevenLabs API error (\(statusCode)): \(message)"
        case .audioDecodingFailed:
            return "Failed to decode audio data from ElevenLabs."
        case .audioSessionError(let message):
            return "Audio session error: \(message)"
        case .voiceNotAvailable(let name):
            return "Voice '\(name)' is not available on this device."
        case .synthesisFailed(let message):
            return "Speech synthesis failed: \(message)"
        }
    }
}

// MARK: - VoiceService

/// Service for text-to-speech using either ElevenLabs or Apple's AVSpeechSynthesizer.
///
/// This service provides a unified interface for TTS that abstracts the underlying
/// provider. It uses AVSpeechSynthesizer as a built-in fallback when ElevenLabs
/// is not configured or unavailable.
///
/// ## Architecture
/// - Uses protocol-based abstraction for testability
/// - AVSpeechSynthesizer for system TTS (always available)
/// - ElevenLabs REST API for high-quality AI voices (requires API key)
/// - Audio playback via AVAudioPlayer for ElevenLabs-generated audio
/// - Combine publishers for reactive state updates
///
/// ## Usage
/// ```swift
/// let service = VoiceService.shared
/// service.updateSettings(VoiceSettings.load())
/// await service.speak("Hello from Claude Code!")
/// ```
final class VoiceService: NSObject, VoiceServiceProtocol {

    // MARK: - Singleton

    static let shared = VoiceService()

    // MARK: - Published State

    private let _playbackState = CurrentValueSubject<VoicePlaybackState, Never>(.idle)
    private let _errors = PassthroughSubject<VoiceServiceError, Never>()

    var playbackState: VoicePlaybackState {
        _playbackState.value
    }

    var playbackStatePublisher: AnyPublisher<VoicePlaybackState, Never> {
        _playbackState.eraseToAnyPublisher()
    }

    var errorPublisher: AnyPublisher<VoiceServiceError, Never> {
        _errors.eraseToAnyPublisher()
    }

    // MARK: - Private Properties

    private var settings: VoiceSettings = .default
    private let synthesizer = AVSpeechSynthesizer()
    private var audioPlayer: AVAudioPlayer?
    private let urlSession: URLSession

    /// Queue of text chunks waiting to be spoken.
    private var speechQueue: [String] = []

    /// Whether we are currently processing the queue.
    private var isProcessingQueue = false

    // MARK: - ElevenLabs Configuration

    private static let elevenLabsBaseURL = "https://api.elevenlabs.io/v1"

    // MARK: - Initialization

    override init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.urlSession = URLSession(configuration: config)

        super.init()
        synthesizer.delegate = self
    }

    /// Initializer for testing with a custom URLSession.
    init(urlSession: URLSession) {
        self.urlSession = urlSession
        super.init()
        synthesizer.delegate = self
    }

    // MARK: - Public Methods

    func speak(_ text: String) async {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return
        }

        // Stop any current speech first
        stop()

        switch settings.provider {
        case .system:
            speakWithSystem(text)
        case .elevenLabs:
            await speakWithElevenLabs(text)
        }
    }

    func stop() {
        speechQueue.removeAll()
        isProcessingQueue = false

        if synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }

        audioPlayer?.stop()
        audioPlayer = nil

        _playbackState.send(.idle)
    }

    func pause() {
        if synthesizer.isSpeaking {
            synthesizer.pauseSpeaking(at: .word)
            _playbackState.send(.paused)
        } else if let player = audioPlayer, player.isPlaying {
            player.pause()
            _playbackState.send(.paused)
        }
    }

    func resume() {
        if synthesizer.isPaused {
            synthesizer.continueSpeaking()
            _playbackState.send(.speaking)
        } else if let player = audioPlayer, !player.isPlaying {
            player.play()
            _playbackState.send(.speaking)
        }
    }

    func updateSettings(_ settings: VoiceSettings) {
        self.settings = settings
    }

    func isProviderAvailable() -> Bool {
        switch settings.provider {
        case .system:
            return true
        case .elevenLabs:
            return KeychainHelper.exists(.elevenLabsApiKey)
        }
    }

    // MARK: - System TTS (AVSpeechSynthesizer)

    /// Speak text using Apple's built-in AVSpeechSynthesizer.
    private func speakWithSystem(_ text: String) {
        configureAudioSession()

        let utterance = AVSpeechUtterance(string: text)

        // Apply settings
        utterance.rate = mapSpeechRate(settings.speechRate)
        utterance.volume = settings.volume

        // Select voice
        if let voiceId = settings.systemVoiceIdentifier,
           let voice = AVSpeechSynthesisVoice(identifier: voiceId) {
            utterance.voice = voice
        } else {
            // Default to the device's preferred language
            utterance.voice = AVSpeechSynthesisVoice(language: Locale.current.language.languageCode?.identifier ?? "en-US")
        }

        _playbackState.send(.speaking)
        synthesizer.speak(utterance)
    }

    /// Map a user-facing speech rate (0.5-2.0) to AVSpeechUtterance rate range.
    ///
    /// AVSpeechUtterance rate ranges from 0.0 to 1.0 where:
    /// - 0.0 = AVSpeechUtteranceMinimumSpeechRate
    /// - 0.5 = AVSpeechUtteranceDefaultSpeechRate
    /// - 1.0 = AVSpeechUtteranceMaximumSpeechRate
    private func mapSpeechRate(_ rate: Float) -> Float {
        // Map 0.5-2.0 user range to 0.3-0.7 AVSpeechUtterance range
        let minRate: Float = AVSpeechUtteranceMinimumSpeechRate
        let maxRate: Float = AVSpeechUtteranceMaximumSpeechRate
        let defaultRate: Float = AVSpeechUtteranceDefaultSpeechRate

        if rate <= 1.0 {
            // Map 0.5-1.0 to minRate-defaultRate
            let normalized = (rate - 0.5) / 0.5
            return minRate + normalized * (defaultRate - minRate)
        } else {
            // Map 1.0-2.0 to defaultRate-maxRate
            let normalized = (rate - 1.0) / 1.0
            return defaultRate + normalized * (maxRate - defaultRate)
        }
    }

    // MARK: - ElevenLabs TTS

    /// Speak text using ElevenLabs API.
    private func speakWithElevenLabs(_ text: String) async {
        guard let apiKey = KeychainHelper.readString(.elevenLabsApiKey) else {
            let error = VoiceServiceError.apiKeyMissing
            _errors.send(error)
            // Fall back to system TTS
            speakWithSystem(text)
            return
        }

        _playbackState.send(.loading)

        do {
            let audioData = try await requestElevenLabsAudio(
                text: text,
                voiceId: settings.elevenLabsVoiceId,
                modelId: settings.elevenLabsModelId,
                apiKey: apiKey
            )

            try await playAudioData(audioData)
        } catch let error as VoiceServiceError {
            _errors.send(error)
            // Fall back to system TTS on ElevenLabs failure
            #if DEBUG
            print("[VoiceService] ElevenLabs failed, falling back to system TTS: \(error)")
            #endif
            speakWithSystem(text)
        } catch {
            let voiceError = VoiceServiceError.synthesisFailed(error.localizedDescription)
            _errors.send(voiceError)
            speakWithSystem(text)
        }
    }

    /// Request audio from ElevenLabs Text-to-Speech API.
    ///
    /// - Parameters:
    ///   - text: The text to convert to speech.
    ///   - voiceId: The ElevenLabs voice identifier.
    ///   - modelId: The ElevenLabs model identifier.
    ///   - apiKey: The ElevenLabs API key.
    /// - Returns: Raw audio data (MP3 format).
    private func requestElevenLabsAudio(
        text: String,
        voiceId: String,
        modelId: String,
        apiKey: String
    ) async throws -> Data {
        let urlString = "\(Self.elevenLabsBaseURL)/text-to-speech/\(voiceId)"
        guard let url = URL(string: urlString) else {
            throw VoiceServiceError.networkError("Invalid URL: \(urlString)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("audio/mpeg", forHTTPHeaderField: "Accept")
        request.setValue(apiKey, forHTTPHeaderField: "xi-api-key")

        let body = ElevenLabsTTSRequest(
            text: text,
            modelId: modelId,
            voiceSettings: ElevenLabsVoiceSettings(
                stability: 0.5,
                similarityBoost: 0.75,
                style: 0.0,
                useSpeakerBoost: true
            )
        )

        request.httpBody = try JSONEncoder().encode(body)

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await urlSession.data(for: request)
        } catch {
            throw VoiceServiceError.networkError(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw VoiceServiceError.networkError("Invalid response")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw VoiceServiceError.apiError(
                statusCode: httpResponse.statusCode,
                message: errorMessage
            )
        }

        guard !data.isEmpty else {
            throw VoiceServiceError.audioDecodingFailed
        }

        return data
    }

    /// Play raw audio data using AVAudioPlayer.
    @MainActor
    private func playAudioData(_ data: Data) throws {
        configureAudioSession()

        do {
            audioPlayer = try AVAudioPlayer(data: data)
            audioPlayer?.delegate = self
            audioPlayer?.volume = settings.volume
            audioPlayer?.prepareToPlay()

            guard audioPlayer?.play() == true else {
                throw VoiceServiceError.synthesisFailed("AVAudioPlayer failed to start playback")
            }

            _playbackState.send(.speaking)
        } catch let error as VoiceServiceError {
            throw error
        } catch {
            throw VoiceServiceError.audioDecodingFailed
        }
    }

    // MARK: - Audio Session

    /// Configure the audio session for speech playback.
    private func configureAudioSession() {
        do {
            let audioSession = AVAudioSession.sharedInstance()
            try audioSession.setCategory(.playback, mode: .spokenAudio, options: [.duckOthers])
            try audioSession.setActive(true)
        } catch {
            #if DEBUG
            print("[VoiceService] Failed to configure audio session: \(error)")
            #endif
            _errors.send(.audioSessionError(error.localizedDescription))
        }
    }

    /// Deactivate the audio session when speech completes.
    private func deactivateAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            #if DEBUG
            print("[VoiceService] Failed to deactivate audio session: \(error)")
            #endif
        }
    }
}

// MARK: - AVSpeechSynthesizerDelegate

extension VoiceService: AVSpeechSynthesizerDelegate {
    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance) {
        _playbackState.send(.speaking)
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        _playbackState.send(.idle)
        deactivateAudioSession()
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance) {
        _playbackState.send(.idle)
        deactivateAudioSession()
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didPause utterance: AVSpeechUtterance) {
        _playbackState.send(.paused)
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didContinue utterance: AVSpeechUtterance) {
        _playbackState.send(.speaking)
    }
}

// MARK: - AVAudioPlayerDelegate

extension VoiceService: AVAudioPlayerDelegate {
    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        audioPlayer = nil
        _playbackState.send(.idle)
        deactivateAudioSession()
    }

    func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        audioPlayer = nil
        _playbackState.send(.idle)
        if let error = error {
            _errors.send(.audioDecodingFailed)
            #if DEBUG
            print("[VoiceService] Audio decode error: \(error)")
            #endif
        }
        deactivateAudioSession()
    }
}

// MARK: - ElevenLabs API Types

/// Request body for ElevenLabs Text-to-Speech API.
struct ElevenLabsTTSRequest: Codable {
    let text: String
    let modelId: String
    let voiceSettings: ElevenLabsVoiceSettings

    enum CodingKeys: String, CodingKey {
        case text
        case modelId = "model_id"
        case voiceSettings = "voice_settings"
    }
}

/// Voice settings for ElevenLabs TTS.
struct ElevenLabsVoiceSettings: Codable {
    let stability: Float
    let similarityBoost: Float
    let style: Float
    let useSpeakerBoost: Bool

    enum CodingKeys: String, CodingKey {
        case stability
        case similarityBoost = "similarity_boost"
        case style
        case useSpeakerBoost = "use_speaker_boost"
    }
}

File: apps/mobile/android/app/src/main/java/com/enflame/happy/data/voice/VoiceService.kt
package com.enflame.happy.data.voice

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import android.os.IBinder
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import com.enflame.happy.data.local.TokenStorage
import com.enflame.happy.domain.model.VoicePlaybackState
import com.enflame.happy.domain.model.VoiceServiceError
import com.enflame.happy.domain.model.VoiceSettings
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.util.Locale
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for text-to-speech using either ElevenLabs or Android's built-in TextToSpeech.
 *
 * This service provides a unified interface for TTS that abstracts the underlying
 * provider. It uses Android's [TextToSpeech] as a built-in fallback when ElevenLabs
 * is not configured or unavailable.
 *
 * ## Architecture
 * - Android [TextToSpeech] for system TTS (always available)
 * - ElevenLabs REST API for high-quality AI voices (requires API key)
 * - [VoicePlaybackService] foreground service for background audio playback (HAP-1021)
 * - [AudioDeviceManager] for Bluetooth and audio output routing (HAP-1021)
 * - Kotlin [StateFlow]/[SharedFlow] for reactive state updates
 *
 * ## Background Audio (HAP-1021)
 * ElevenLabs audio playback is delegated to [VoicePlaybackService], a foreground
 * service that:
 * - Continues playback when the app is backgrounded
 * - Shows a MediaStyle notification with play/pause/stop controls
 * - Integrates with system media controls and MediaSession
 * - Handles audio focus changes (duck/pause for interruptions)
 *
 * ## Usage
 * ```kotlin
 * val service = voiceService // injected via Hilt
 * service.updateSettings(settings)
 * service.speak("Hello from Claude Code!")
 * ```
 */
@Singleton
class VoiceService @Inject constructor(
    @ApplicationContext private val context: Context,
    private val elevenLabsApiService: ElevenLabsApiService,
    private val tokenStorage: TokenStorage,
    private val audioDeviceManager: AudioDeviceManager
) {

    // --- State ---

    private val _playbackState = MutableStateFlow(VoicePlaybackState.IDLE)

    /** Current playback state as a reactive flow. */
    val playbackState: StateFlow<VoicePlaybackState> = _playbackState.asStateFlow()

    private val _errors = MutableSharedFlow<VoiceServiceError>(extraBufferCapacity = 1)

    /** Errors emitted during voice operations. */
    val errors: SharedFlow<VoiceServiceError> = _errors.asSharedFlow()

    // --- Private Properties ---

    private var settings: VoiceSettings = VoiceSettings()
    private var textToSpeech: TextToSpeech? = null
    private var isTtsInitialized = false
    private val audioManager: AudioManager =
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private var audioFocusRequest: AudioFocusRequest? = null

    /** Bound foreground service for background audio playback. */
    private var playbackService: VoicePlaybackService? = null
    private var isServiceBound = false

    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    /**
     * ServiceConnection for binding to [VoicePlaybackService].
     *
     * When bound, observes the service's playback state and forwards
     * it to this service's state flow.
     */
    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            val binder = service as VoicePlaybackService.VoicePlaybackBinder
            playbackService = binder.getService().also { svc ->
                svc.volume = settings.volume
                svc.onPlaybackComplete = {
                    _playbackState.value = VoicePlaybackState.IDLE
                }
                svc.onPlaybackError = { message ->
                    _playbackState.value = VoicePlaybackState.IDLE
                    _errors.tryEmit(VoiceServiceError.SynthesisFailed(message))
                }

                // Observe foreground service playback state
                serviceScope.launch {
                    svc.playbackState.collect { state ->
                        _playbackState.value = state
                    }
                }
            }
            isServiceBound = true
            Log.d(TAG, "VoicePlaybackService bound")
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            playbackService = null
            isServiceBound = false
            Log.d(TAG, "VoicePlaybackService disconnected")
        }
    }

    companion object {
        private const val TAG = "VoiceService"
        private const val UTTERANCE_ID = "happy_voice_utterance"
        private const val KEY_ELEVEN_LABS_API_KEY = "eleven_labs_api_key"
    }

    init {
        initializeSystemTts()
        bindPlaybackService()
    }

    // --- Public Methods ---

    /**
     * Speak the given text aloud using the configured provider.
     *
     * If the text is blank, this method does nothing. Any currently playing
     * speech is stopped before new speech begins.
     *
     * @param text The text to convert to speech.
     */
    suspend fun speak(text: String) {
        if (text.isBlank()) return

        stop()

        when (settings.provider) {
            com.enflame.happy.domain.model.VoiceProvider.SYSTEM -> speakWithSystem(text)
            com.enflame.happy.domain.model.VoiceProvider.ELEVEN_LABS -> speakWithElevenLabs(text)
        }
    }

    /**
     * Stop any current speech playback.
     */
    fun stop() {
        textToSpeech?.stop()

        // Stop the foreground playback service
        if (isServiceBound) {
            playbackService?.stopPlayback()
        } else {
            // Send a stop intent even if not bound
            try {
                context.startService(VoicePlaybackService.createStopIntent(context))
            } catch (e: Exception) {
                Log.w(TAG, "Failed to send stop intent to playback service", e)
            }
        }

        abandonAudioFocus()
        _playbackState.value = VoicePlaybackState.IDLE
    }

    /**
     * Pause the current speech playback.
     *
     * Only ElevenLabs audio (via foreground service) supports pause. System TTS
     * does not support pausing mid-utterance on Android.
     */
    fun pause() {
        playbackService?.pause()
    }

    /**
     * Resume paused speech playback.
     */
    fun resume() {
        playbackService?.resume()
    }

    /**
     * Update the voice settings.
     *
     * @param newSettings The updated voice settings.
     */
    fun updateSettings(newSettings: VoiceSettings) {
        settings = newSettings
        playbackService?.volume = newSettings.volume
    }

    /**
     * Check if the current provider is available and configured.
     *
     * @return `true` if the provider is ready for use.
     */
    fun isProviderAvailable(): Boolean {
        return when (settings.provider) {
            com.enflame.happy.domain.model.VoiceProvider.SYSTEM -> isTtsInitialized
            com.enflame.happy.domain.model.VoiceProvider.ELEVEN_LABS ->
                tokenStorage.exists(KEY_ELEVEN_LABS_API_KEY)
        }
    }

    /**
     * Save the ElevenLabs API key to secure storage.
     *
     * @param apiKey The API key to store.
     */
    fun saveApiKey(apiKey: String) {
        tokenStorage.saveString(KEY_ELEVEN_LABS_API_KEY, apiKey)
    }

    /**
     * Check if an ElevenLabs API key is stored.
     *
     * @return `true` if an API key exists in secure storage.
     */
    fun hasApiKey(): Boolean {
        return tokenStorage.exists(KEY_ELEVEN_LABS_API_KEY)
    }

    /**
     * Delete the stored ElevenLabs API key.
     */
    fun deleteApiKey() {
        tokenStorage.remove(KEY_ELEVEN_LABS_API_KEY)
    }

    /**
     * Release all resources held by this service.
     *
     * Should be called when the service is no longer needed.
     */
    fun release() {
        stop()
        textToSpeech?.shutdown()
        textToSpeech = null
        isTtsInitialized = false
        unbindPlaybackService()
        audioDeviceManager.stopMonitoring()
    }

    // --- Playback Service Binding ---

    /**
     * Bind to the [VoicePlaybackService] for background audio control.
     */
    private fun bindPlaybackService() {
        try {
            val intent = Intent(context, VoicePlaybackService::class.java)
            context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to bind VoicePlaybackService", e)
        }
    }

    /**
     * Unbind from the [VoicePlaybackService].
     */
    private fun unbindPlaybackService() {
        if (isServiceBound) {
            try {
                context.unbindService(serviceConnection)
            } catch (e: Exception) {
                Log.w(TAG, "Failed to unbind VoicePlaybackService", e)
            }
            isServiceBound = false
            playbackService = null
        }
    }

    // --- System TTS ---

    /**
     * Initialize the Android TextToSpeech engine.
     */
    private fun initializeSystemTts() {
        textToSpeech = TextToSpeech(context) { status ->
            isTtsInitialized = status == TextToSpeech.SUCCESS
            if (status != TextToSpeech.SUCCESS) {
                Log.e(TAG, "TextToSpeech initialization failed with status: $status")
            }
        }
    }

    /**
     * Speak text using Android's built-in TextToSpeech.
     */
    private fun speakWithSystem(text: String) {
        val tts = textToSpeech ?: run {
            _errors.tryEmit(VoiceServiceError.SynthesisFailed("TextToSpeech not initialized"))
            return
        }

        if (!isTtsInitialized) {
            _errors.tryEmit(VoiceServiceError.SynthesisFailed("TextToSpeech not ready"))
            return
        }

        if (!requestAudioFocus()) {
            _errors.tryEmit(VoiceServiceError.AudioFocusError("Could not acquire audio focus"))
            return
        }

        // Apply settings
        tts.setSpeechRate(settings.speechRate)

        // Set voice if specified
        settings.systemVoiceName?.let { voiceName ->
            val availableVoices = tts.voices ?: emptySet()
            val selectedVoice = availableVoices.find { it.name == voiceName }
            if (selectedVoice != null) {
                tts.voice = selectedVoice
            } else {
                tts.language = Locale.getDefault()
            }
        } ?: run {
            tts.language = Locale.getDefault()
        }

        // Set up listener
        tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
            override fun onStart(utteranceId: String?) {
                _playbackState.value = VoicePlaybackState.SPEAKING
            }

            override fun onDone(utteranceId: String?) {
                _playbackState.value = VoicePlaybackState.IDLE
                abandonAudioFocus()
            }

            @Deprecated("Deprecated in Java", ReplaceWith("onError(utteranceId, errorCode)"))
            override fun onError(utteranceId: String?) {
                _playbackState.value = VoicePlaybackState.IDLE
                _errors.tryEmit(VoiceServiceError.SynthesisFailed("TTS error"))
                abandonAudioFocus()
            }

            override fun onError(utteranceId: String?, errorCode: Int) {
                _playbackState.value = VoicePlaybackState.IDLE
                _errors.tryEmit(VoiceServiceError.SynthesisFailed("TTS error code: $errorCode"))
                abandonAudioFocus()
            }
        })

        _playbackState.value = VoicePlaybackState.SPEAKING
        tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, UTTERANCE_ID)
    }

    // --- ElevenLabs TTS ---

    /**
     * Speak text using ElevenLabs API.
     *
     * Falls back to system TTS if the API call fails.
     * Audio playback is delegated to [VoicePlaybackService] for background support.
     */
    private suspend fun speakWithElevenLabs(text: String) {
        val apiKey = tokenStorage.readString(KEY_ELEVEN_LABS_API_KEY)
        if (apiKey == null) {
            _errors.tryEmit(VoiceServiceError.ApiKeyMissing)
            // Fall back to system TTS
            speakWithSystem(text)
            return
        }

        _playbackState.value = VoicePlaybackState.LOADING

        try {
            val audioData = requestElevenLabsAudio(
                text = text,
                voiceId = settings.elevenLabsVoiceId,
                modelId = settings.elevenLabsModelId,
                apiKey = apiKey
            )

            playAudioViaForegroundService(audioData)
        } catch (e: VoiceServiceError) {
            _errors.tryEmit(e)
            Log.w(TAG, "ElevenLabs failed, falling back to system TTS: ${e.message}")
            speakWithSystem(text)
        } catch (e: Exception) {
            val error = VoiceServiceError.SynthesisFailed(e.localizedMessage ?: "Unknown error")
            _errors.tryEmit(error)
            Log.w(TAG, "ElevenLabs failed, falling back to system TTS", e)
            speakWithSystem(text)
        }
    }

    /**
     * Request audio from ElevenLabs Text-to-Speech API.
     *
     * @param text The text to convert to speech.
     * @param voiceId The ElevenLabs voice identifier.
     * @param modelId The ElevenLabs model identifier.
     * @param apiKey The ElevenLabs API key.
     * @return Raw audio data (MP3 format).
     */
    private suspend fun requestElevenLabsAudio(
        text: String,
        voiceId: String,
        modelId: String,
        apiKey: String
    ): ByteArray = withContext(Dispatchers.IO) {
        val request = ElevenLabsTtsRequest(
            text = text,
            modelId = modelId,
            voiceSettings = ElevenLabsVoiceSettingsDto(
                stability = 0.5f,
                similarityBoost = 0.75f,
                style = 0.0f,
                useSpeakerBoost = true
            )
        )

        try {
            val response = elevenLabsApiService.textToSpeech(
                voiceId = voiceId,
                apiKey = apiKey,
                request = request
            )

            if (!response.isSuccessful) {
                val errorBody = response.errorBody()?.string() ?: "Unknown error"
                throw VoiceServiceError.ApiError(
                    statusCode = response.code(),
                    detail = errorBody
                )
            }

            val body = response.body()
                ?: throw VoiceServiceError.AudioDecodingFailed

            val bytes = body.bytes()
            if (bytes.isEmpty()) {
                throw VoiceServiceError.AudioDecodingFailed
            }

            bytes
        } catch (e: VoiceServiceError) {
            throw e
        } catch (e: Exception) {
            throw VoiceServiceError.NetworkError(e.localizedMessage ?: "Connection failed")
        }
    }

    /**
     * Play audio data via the [VoicePlaybackService] foreground service.
     *
     * Writes audio bytes to a temporary file and starts the foreground service
     * for background-capable playback with media notification.
     *
     * @param data The raw audio bytes (MP3 format).
     */
    private suspend fun playAudioViaForegroundService(data: ByteArray) {
        withContext(Dispatchers.IO) {
            // Write audio data to a temp file for the foreground service
            val tempFile = File.createTempFile("happy_tts_", ".mp3", context.cacheDir)
            FileOutputStream(tempFile).use { it.write(data) }

            withContext(Dispatchers.Main) {
                try {
                    // Start the foreground service with the audio file
                    val intent = VoicePlaybackService.createPlayIntent(
                        context = context,
                        audioFilePath = tempFile.absolutePath,
                        title = "Happy Voice"
                    )

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        context.startForegroundService(intent)
                    } else {
                        context.startService(intent)
                    }

                    // If bound, also trigger playback directly
                    playbackService?.let { service ->
                        service.volume = settings.volume
                        service.playAudio(tempFile.absolutePath, "Happy Voice")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to start foreground playback service", e)
                    _playbackState.value = VoicePlaybackState.IDLE
                    _errors.tryEmit(VoiceServiceError.SynthesisFailed(
                        e.localizedMessage ?: "Failed to start playback"
                    ))
                    tempFile.delete()
                }
            }
        }
    }

    // --- Audio Focus (for System TTS only) ---

    /**
     * Request audio focus for system TTS speech playback.
     *
     * ElevenLabs audio focus is managed by [VoicePlaybackService].
     *
     * @return `true` if audio focus was granted.
     */
    private fun requestAudioFocus(): Boolean {
        val focusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .build()
            )
            .setOnAudioFocusChangeListener { focusChange ->
                when (focusChange) {
                    AudioManager.AUDIOFOCUS_LOSS,
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                        textToSpeech?.stop()
                        _playbackState.value = VoicePlaybackState.IDLE
                        abandonAudioFocus()
                    }
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                        // Lower volume temporarily (handled by system ducking)
                    }
                }
            }
            .build()

        audioFocusRequest = focusRequest
        val result = audioManager.requestAudioFocus(focusRequest)
        return result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED
    }

    /**
     * Abandon audio focus after speech completes.
     */
    private fun abandonAudioFocus() {
        audioFocusRequest?.let { request ->
            audioManager.abandonAudioFocusRequest(request)
            audioFocusRequest = null
        }
    }
}
